File: internal/payload/payload.go
package payload

import (
	"fmt"
	"strings"

	"sisyphus/internal/ado"
)

const baseTemplate = "Instructions:\n" +
	"- You are attempting to resolve a failing build located at {failing_build_url}\n\n" +
	"Here is a log exerpt from the failing build:\n\n" +
	"```\n" +
	"{logExcerpt}\n" +
	"```\n\n" +
	"- The build definition is based upon the yaml at '{build_definition_yaml_path}'.\n" +
	"- Make any local changes necessary that you feel will address the failure.\n" +
	"- If absolutely necessary, ADO_PAT is available in environment if you need to make limited requests of ADO with an identity that can interrogate the build system over REST calls. Do not abuse this." +
	"- State your conclusions, then exit this copilot prompt with 0 if you feel you succeeded, otherwise exit with 1\n"

func BuildFailureInstructions(buildYAMLPath string, repoPath string, buildID *int, client *ado.Client, logMaxBytes int, failureDetail string) (string, error) {
	_ = repoPath
	buildURL := "<queue submission failed>"
	logExcerpt := ""
	failureDetail = strings.TrimSpace(failureDetail)
	if buildID != nil {
		buildURL = fmt.Sprintf(
			"%s/%s/%s/_build/results?buildId=%d&view=results",
			strings.TrimRight(client.BaseURL, "/"),
			client.Org,
			client.Project,
			*buildID,
		)
		fetched, err := ado.FetchFailureExcerpt(client, *buildID, logMaxBytes)
		if err != nil {
			if failureDetail == "" {
				return "", err
			}
			logExcerpt = failureDetail
			failureDetail = fmt.Sprintf("Failed to fetch build logs: %v", err)
		} else {
			logExcerpt = fetched
		}
	} else {
		logExcerpt = failureDetail
		failureDetail = ""
	}
	if failureDetail != "" && logExcerpt != "" {
		logExcerpt = logExcerpt + "\n\nAdditional failure detail:\n" + failureDetail
	}
	if logExcerpt == "" {
		logExcerpt = "<no failure details available>"
	}
	yamlPath := buildYAMLPath
	if yamlPath == "" {
		yamlPath = "<unknown yaml path>"
	}

	out := strings.NewReplacer(
		"{failing_build_url}", buildURL,
		"{logExcerpt}", logExcerpt,
		"{build_definition_yaml_path}", yamlPath,
	).Replace(baseTemplate)
	return out, nil
}

File: agent/payload.py
from __future__ import annotations

from pathlib import Path

from agent.ado import logs
from agent.ado.client import ADOClient


BASE_TEMPLATE = """
Issue: {issue}

Instructions:
- We are attempting to resolve a failing build at
"""


FAILURE_TEMPLATE = """
Build Failure Detected
Build ID: {build_id}

Failed Log Excerpt (truncated):
{log_excerpt}
"""


def build_initial_instructions(*, issue: str, build_def: str, repo_path: Path) -> str:
    return BASE_TEMPLATE.format(
        issue=issue,
        build_def=build_def,
        repo_path=repo_path,
        instructions_path=repo_path / "instructions.md",
    )


def build_failure_instructions(
    *,
    issue: str,
    build_def: str,
    repo_path: Path,
    build_id: int,
    client: ADOClient,
    log_max_bytes: int,
) -> str:
    base = build_initial_instructions(issue=issue, build_def=build_def, repo_path=repo_path)
    log_excerpt = logs.fetch_failure_excerpt(client, build_id, max_bytes=log_max_bytes)
    return base + FAILURE_TEMPLATE.format(build_id=build_id, log_excerpt=log_excerpt)


def write_instructions(path: Path, content: str) -> None:
    path.write_text(content, encoding="utf-8")

